### ISSUE TITLE
Architecture Improvements

### ISSUE THREAD

**Issue by person594** [uber/deck.gl#270]
Ideas for medium- to long-term changes to the architecture of deck.gl.  This list will certainly see modifications and additions.

- Increase separation between model and view.  Layers are currently responsible for both their underlying data and for the display of that data. Perhaps add something like a scene graph to manage rendering in a more centralized way, and have layers add to or modify that scene graph.

- Finalize how effects will be managed and rendered.  Effects are currently responsible for their own rendering much like layers -- if a scene graph is added, effects could similarly modify the scene graph instead of doing rendering themselves.  Once more effects are implemented, API decisions regarding them should be more clear.

**Comment by person594** [comment]
@shaojingli

**Comment by shaojingli** [comment]
Thanks @person594 . These are very valuable feedback. If you have any other architectural suggestions, just put it here and we can discuss and plan for it accordingly.

**Comment by ibgreen** [comment]
@person594  These are important questions, thanks for getting the discussion started:

* I understand that when adding effects it would be helpful to have the framework to take more control of the shaders, to be able to add effects support in a single place.
* And there is no doubt that we need to organize our shader code as the library grows.

Still, I feel that that going to a scenegraph in particular is a big decision, as deck.gl was created as a functional/reactive library with the goal to make it easy to work directly with GLSL to create unique shader based visualizations, using e.g. instanced rendering and other techniques such as WebGL extensions and WebGL2 etc.

FWIW, in deck.gl v2 we had a scenegraph (the luma.gl `Scene` class) and the layer manager added all the the layer models to the scene before rendering. But it seemed like a poor fit for the "functional" (reactive) ambitions of these libraries, so in v3 scene graph support was removed, completely from deck.gl and deprecated in luma.gl.

As a possible alternative to scenegraphs, there is ongoing work on extensions to the `assembleShaders` system that would essentially allow dynamic injection of shader code into layers - I suspect it could provide some of the functionality that our effects would need from a scene graph. I'll try to get that PR up soon so that this can be reviewed/discussed.

Because of this hesitation, it would be helpful if we can be more precise in our requirements on the architecture, i.e. what actual limitations are we running up against when implementing "effect X" or "layer Y" i.e.:
* To implement shadow map effect, I would need to inject code X into every vertex shader.
* To implement reflections correctly/efficiently, I need the following changes to the Viewport...
* To implement blur effect, I would need the following in fragment shaders...
etc.

Then we could efficiently evaluate any planned changes to current architecture and see if they are enough, or if e.g. a scenegraph is truly needed.

I will also spend some time documenting ongoing architectural work and ideas to make sure everyone can get up to speed.

**Comment by shaojingli** [comment]
I agree with @ibgreen that using a centralized scene graph would be a digress from our goal to make the library functional/reactive and, also, a scene graph like what we see in most game engines sounds like a huge overkill for our purposes - we don't have that complicated a scene for geo data viz. However I also do understand @person594's desire 1) to have a more centralized way of controlling rendering pipeline and 2) enable resource sharing (attribute buffer sharing, program object sharing...etc...) across layers for reducing code replication and increasing rendering and data transferring efficiency. This makes stuff like EffectsManager (which only deals with rendering not data) easier to be integrated with existing code base  and has also been mentioned by @abmai and other deck.gl users. 

I would say that for now, we should keep what I called "democratic" architecture of deck.gl and we could improve it in the future if use cases compel us to do so.

**Comment by person594** [comment]
@ibgreen That makes sense -- from my interactions with the code, a scene graph seemed like a good direction to go, but given the narrow scope of what I've been working on, I don't have much perspective on how that would work with the rest of the codebase.

Code injection/modification would certainly fix one of the major problems I saw a scene graph as a solution to.  As I'll be working on lighting after christmas, it should quickly become quite clear what I'll need from assembleshaders.  I foresee some subtle conceptual problems (i.e. how do we specify that the screen grid layer shouldn't cast a shadow?), but if we have a general way to modify shaders on the fly that should definitely be enough to implement lighting, distance-dependent blur, and really anything I can think of.

The other main gripe that made me feel a scene grid might be worthwhile was that I had to use the LayerManager in order to even render the scene, while it feels that I shouldn't even have to know what a layer is in order to just draw everything (it feels the process of rendering should be abstracted away from knowledge of what exactly is being rendered).  This is actually a really minor complaint now that I think about it, and if it continues to bother me it would be much more easily addressed by adding another layer of indirection to the rendering process (e.g. a function to render everything that doesn't have the word layer in its name) that it would by introducing a scene graph.
